File: src/main/java/supermetroid/trackable/BeamTrackable.java
package supermetroid.trackable;

import supermetroid.State;

import javax.imageio.ImageIO;
import javax.swing.*;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class BeamTrackable extends Trackable {
  private String current;
  private Map<String, ImageIcon> beams = new HashMap<>();

  public BeamTrackable() throws Exception {
    for(var c = 0; c < 2; c++) {
      for (var w = 0; w < 2; w++) {
        for (var i = 0; i < 2; i++) {
          for (var s = 0; s < 2; s++) {
            for (var p = 0; p < 2; p++) {
              String key = (c == 1 ? "a" : "i") + (w == 1 ? "a" : "i") +(i == 1 ? "a" : "i") + (s == 1 ? "a" : "i") + (p == 1 ? "a" : "i");
              String image = "/beams/cwisp-" + key + ".png";
              ImageIcon icon = new ImageIcon(ImageIO.read(Objects.requireNonNull(BooleanTrackable.class.getResourceAsStream(image))));
              beams.put(key, icon);
            }
          }
        }
      }
    }
  }

  @Override
  public void setIcon(State state) {
    boolean c = (state.charge & 0x10) > 0;
    boolean w = (state.beams & 0x1) > 0;
    boolean i = (state.beams & 0x2) > 0;
    boolean s = (state.beams & 0x4) > 0;
    boolean p = (state.beams & 0x8) > 0;
    String key = (c ? "a" : "i") + (w ? "a" : "i") +(i ? "a" : "i") + (s ? "a" : "i") + (p ? "a" : "i");

    if (!key.equals(current)) {
      current = key;
      label.setIcon(beams.get(key));
    }
  }

  @Override
  public void setDefault() {
    label.setIcon(beams.get("iiiii"));
  }
}

File: src/main/java/supermetroid/trackable/BeamTrackable.kt
package supermetroid.trackable

import supermetroid.State
import supermetroid.readResourceToIcon
import java.util.*
import javax.swing.ImageIcon

class BeamTrackable : Trackable() {
    private var current: String? = null
    private val beams: MutableMap<String, ImageIcon> = HashMap()

    init {
        for (combinations in 0 until 32) {
            val c = if ((combinations and 0x1) > 0) "a" else "i"
            val w = if ((combinations and 0x2) > 0) "a" else "i"
            val i = if ((combinations and 0x4) > 0) "a" else "i"
            val s = if ((combinations and 0x8) > 0) "a" else "i"
            val p = if ((combinations and 0x10) > 0) "a" else "i"

            val key = "$c$w$i$s$p"
            beams[key] = readResourceToIcon("/beams/cwisp-$key.png")
        }
    }

    override fun setIcon(state: State) {
        val c = if ((state.charge and 0x10) > 0) "a" else "i"
        val w = if ((state.beams and 0x1) > 0) "a" else "i"
        val i = if ((state.beams and 0x2) > 0) "a" else "i"
        val s = if ((state.beams and 0x4) > 0) "a" else "i"
        val p = if ((state.beams and 0x8) > 0) "a" else "i"

        val key = "$c$w$i$s$p"

        if (key != current) {
            current = key
            label.setIcon(beams[key])
        }
    }

    override fun setDefault() = label.setIcon(beams["iiiii"])
}
